hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestBulkLoadReplication.java

```diff
@@ -0,0 +1,237 @@
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestBulkLoadReplication.class);
+
+  protected static final Logger LOG =
+    LoggerFactory.getLogger(TestBulkLoadReplication.class);
+
+  private static final String SRC_REP_CLUSTER_ID = "src";
+  private static final String DEST_REP_CLUSTER_ID = "dest";
+  private static final String PEER_ID = "1";
+  private static final AtomicInteger BULK_LOADS_COUNT = new AtomicInteger(0);
+
+  @Rule
+  public TestName name = new TestName();
+
+  @ClassRule
+  public static TemporaryFolder testFolder = new TemporaryFolder();
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    setupBulkLoadConfigsForCluster(TestReplicationBase.CONF1, SRC_REP_CLUSTER_ID);
+    setupBulkLoadConfigsForCluster(TestReplicationBase.CONF2, DEST_REP_CLUSTER_ID);
+    TestReplicationBase.setUpBeforeClass();
+  }
+
+  @Before
+  @Override
+  public void setUpBase() throws Exception {
+    super.setUpBase();
+    ReplicationPeerConfig peerConfig = ReplicationPeerConfig.newBuilder()
+      .setClusterKey(UTIL1.getClusterKey()).setSerial(isSerialPeer()).build();
+    UTIL2.getAdmin().addReplicationPeer(PEER_ID, peerConfig);
+    setupCoprocessor(UTIL1);
+    setupCoprocessor(UTIL2);
+  }
+
+  private void setupCoprocessor(HBaseTestingUtility cluster){
+    cluster.getHBaseCluster().getRegions(tableName).forEach(r -> {
+      try {
+        r.getCoprocessorHost()
+          .load(TestBulkLoadReplication.BulkReplicationTestObserver.class, 0,
+            cluster.getConfiguration());
+      } catch (Exception e){
+        LOG.error(e.getMessage(), e);
+      }
+    });
+  }
+
+  @After
+  @Override
+  public void tearDownBase() throws Exception {
+    super.tearDownBase();
+    UTIL2.getAdmin().removeReplicationPeer(PEER_ID);
+  }
+
+  private static void setupBulkLoadConfigsForCluster(Configuration config,
+    String clusterReplicationId) throws Exception {
+    config.setBoolean(HConstants.REPLICATION_BULKLOAD_ENABLE_KEY, true);
+    config.set(REPLICATION_CLUSTER_ID, clusterReplicationId);
+    File sourceConfigFolder = testFolder.newFolder(clusterReplicationId);
+    File sourceConfigFile = new File(sourceConfigFolder.getAbsolutePath()
+      + "/hbase-site.xml");
+    config.writeXml(new FileOutputStream(sourceConfigFile));
+    config.set(REPLICATION_CONF_DIR, testFolder.getRoot().getAbsolutePath());
+  }
+
+  @Test
+  public void testBulkLoadReplicationActiveActive() throws Exception {
+    Table srcTestTable = UTIL1.getConnection().getTable(TestReplicationBase.tableName);
+    Table destTestTable = UTIL2.getConnection().getTable(TestReplicationBase.tableName);
+    byte[] row = Bytes.toBytes("001");
+    byte[] value = Bytes.toBytes("v1");
+    bulkLoadOnCluster(row, value, UTIL1);
+    Thread.sleep(400);
+    assertTableHasValue(srcTestTable, row, value);
+    Thread.sleep(400);
+    assertTableHasValue(destTestTable, row, value);
+    Thread.sleep(400);
```

Comment:
The 400 is time for replication to run? Will this make test a little flakey?

Reply:
Unfortunately, yeah. Not only replication, but bulkloading as well works in an async fasion, so we need to wait the test thread to give a chance for initial bulk load to complete, then wait again for bulk load replication from source to target to complete, then this 3rd additional wait is for getting a chance for detecting a possible infinite loop condition (that we are trying to fix here).

Reply:
Had replaced these sleeps by a CountDownLatch which is updated by a _postBulkLoadHFile_ CP method in this test. I believe this helps preventing potential flakyness, as we can set a huge timeout in the latch.await, without worrying too much with penalising fast runs. Let me know if this looks better.